### PR TITLE
Add then(callback)->Future API to ivalue::Future

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37311 Add then(callback)->Future API to utils/Future.h**

Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

### CODE REVIEW

File: torch/testing/_internal/distributed/rpc/rpc_test.py
Comment (pritamdamania87) [inline torch/testing/_internal/distributed/rpc/rpc_test.py:2371]: Can we add a few more test cases:

1) Test that the callback can be a torchscript function.
2) Attach python callback to future returned by:
     i) rpc_async within torchscript.
     ii) rpc_async calling torchscript function.
3) Add tests to validate that we receive appropriate exceptions if callback doesn't have appropriate signature.

File: torch/csrc/distributed/rpc/init.cpp
Comment (pritamdamania87) [inline torch/csrc/distributed/rpc/init.cpp:365]: Should we have some validation here to ensure that cb is valid? For example its not None and it has the correct signature?

**Review by pritamdamania87 — CHANGES_REQUESTED** [review]

File: torch/csrc/distributed/rpc/init.cpp
Comment (mrshenli) [reply torch/csrc/distributed/rpc/init.cpp:28]: Will rebase when #35154 is landed to see if we still need this.

File: torch/csrc/utils/future.h
Comment (mrshenli) [inline torch/csrc/utils/future.h:137]: I am not sure if this is the appropriate API. Since the returned Future represents the completion of the callback, it should also hold the returned value of the callback?

File: torch/csrc/distributed/rpc/init.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/rpc/init.cpp:381]: nit: I think the `const FutureIValue& fut` shadows the self argument `FutureIValue& fut` on 367?

File: torch/csrc/utils/future.h
Comment (rohan-varma) [reply torch/csrc/utils/future.h:137]: What's the reason for the CB taking in a future instead of the completed value of the future? Is it so that we can access errors on the future when running the callback?

File: torch/csrc/distributed/rpc/init.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/rpc/init.cpp:392]: Thanks for taking care of the error handling! This is great, though I was wondering what you think about making the python user's cb signature something like:
```
def callback(future_result, future_error):
    # User's responsibility to check and handle error
```

with this, we could avoid passing in the future here and instead pass in the result and optional error.

Of course, this is also fine, and we can revisit if user code needs custom error handling.

File: torch/csrc/utils/future.h
Comment (mrshenli) [reply torch/csrc/utils/future.h:137]: > what's the reason for the CB taking in a future instead of the completed value of the future? Is it so that we can access errors on the future when running the callback (if so see above comment)?

Yep, the return value `T` might not be available if the callback crashed.

File: torch/csrc/distributed/rpc/init.cpp
Comment (mrshenli) [reply torch/csrc/distributed/rpc/init.cpp:381]: ahh, right, let me change it to a different name.

File: torch/csrc/distributed/rpc/init.cpp
Comment (mrshenli) [reply torch/csrc/distributed/rpc/init.cpp:392]: > For example, there may be cases where the user does not want to throw if the parent future failed.

I see, this makes sense. In this case, should we let the Python callback API also takes the original future instead of the return value? Otherwise, the two args will be mutual exclusive optional return value and optional error?

Comment (rohan-varma) [reply torch/csrc/distributed/rpc/init.cpp:392]: This would be good too, although it can be done as a follow up too since I don't feel there's a need to block this diff on that. The API is private for now as well.

File: torch/testing/_internal/distributed/rpc/rpc_test.py
Comment (rohan-varma) [inline torch/testing/_internal/distributed/rpc/rpc_test.py:2389]: Is this being here twice intentional? If so, what's it testing?

File: torch/testing/_internal/distributed/rpc/rpc_test.py
Comment (rohan-varma) [inline torch/testing/_internal/distributed/rpc/rpc_test.py:2448]: Nice! Is there a test where a user-defined cb runs on the remote side, throws, and we ensure the right error is propagated back to the client? (I saw there is an error check test above where the cb is local)

**Review by rohan-varma — APPROVED** [review]
Thanks for getting this together so quickly! Looks good to me % some comments. Also it looks like @pritamdamania87 has some comments as well

File: torch/testing/_internal/distributed/rpc/rpc_test.py
Comment (mrshenli) [reply torch/testing/_internal/distributed/rpc/rpc_test.py:2389]: This is testing the cb didn't mess up with the value in the Future. Let me add a comment

File: torch/testing/_internal/distributed/rpc/rpc_test.py
Comment (mrshenli) [reply torch/testing/_internal/distributed/rpc/rpc_test.py:2448]: Let me add this test as well.

File: torch/testing/_internal/distributed/rpc/rpc_test.py
Comment (mrshenli) [reply torch/testing/_internal/distributed/rpc/rpc_test.py]: Cannot test calling `_then` in a TorchScript function yet, as that would require making `_then` torchscrptable.

File: torch/csrc/jit/python/pybind_utils.h
Comment (rohan-varma) [inline torch/csrc/jit/python/pybind_utils.h:115]: Do we need to do any validation on `func_` having the right signature or is the error message when the args are incorrect okay?

Comment (mrshenli) [reply torch/csrc/jit/python/pybind_utils.h:115]: pybind covers this, there is a `test_callback_wrong_args` test to make sure the error message is clear.

**Comment by rohan-varma** [comment]
LGTM, did some testing and verified that I can use this to fix the profiler test issue.

File: torch/csrc/jit/python/init.cpp
Comment (pritamdamania87) [inline torch/csrc/jit/python/init.cpp:834]: Is RPC and IValue future already merged? :)

File: torch/csrc/jit/python/pybind_utils.h
Comment (pritamdamania87) [inline torch/csrc/jit/python/pybind_utils.h:133]: Do we need to catch and rethrow this? Can't we just avoid catching in the first place? That would also preserve the original exception.

File: torch/csrc/jit/python/pybind_utils.h
Comment (pritamdamania87) [inline torch/csrc/jit/python/pybind_utils.h:132]: Shouldn't we mark the future that we're returning as part of `then` with these errors instead of just rethrowing them?

File: torch/csrc/jit/python/init.cpp
Comment (pritamdamania87) [inline torch/csrc/jit/python/init.cpp:833]: Why `_then` instead of just `then`? Is this some convention in JIT?

File: torch/testing/_internal/distributed/rpc/rpc_test.py
Comment (pritamdamania87) [inline torch/testing/_internal/distributed/rpc/rpc_test.py:2458]: Can we have a test similar to this where the remote function is a torchscript function?

**Review by pritamdamania87 — COMMENTED** [review]
Have a few comments inline. Also, was wondering if we should get a review from @wanchaol for the JIT changes?

File: torch/csrc/jit/python/init.cpp
Comment (mrshenli) [reply torch/csrc/jit/python/init.cpp:834]: Yes, they are. See #35154

File: torch/csrc/jit/python/pybind_utils.h
Comment (mrshenli) [reply torch/csrc/jit/python/pybind_utils.h:132]: `ivalue::Future::then(cb)` did that.

File: torch/csrc/jit/python/init.cpp
Comment (mrshenli) [reply torch/csrc/jit/python/init.cpp:833]: No, not sure if the current signature is appropriate, but need to get this in earlier to unblock fixing profiler tests. So mark it `_then` with no API doc for now, and will create a PR on top to expose this API publicly.

File: torch/testing/_internal/distributed/rpc/rpc_test.py
Comment (mrshenli) [reply torch/testing/_internal/distributed/rpc/rpc_test.py:2458]: sure adding.

File: torch/csrc/jit/python/pybind_utils.h
Comment (wanchaol) [inline torch/csrc/jit/python/pybind_utils.h:115]: does this mean that the py::function passed in should always require the user to pass a  future? shall we document this?

**Review by pritamdamania87 — APPROVED** [review]

File: torch/csrc/jit/python/pybind_utils.h
Comment (mrshenli) [reply torch/csrc/jit/python/pybind_utils.h:115]: Yes, let me add a comment here

**Review by wanchaol — APPROVED** [review]
looks good to me.

**Pull request closed without merging**
